src/diff_walker.rs

```diff
@@ -8,13 +8,25 @@ use serde_json::Value;
 
 use crate::{Change, ChangeKind, Error, JsonSchemaType, Range};
 
-pub struct DiffWalker {
-    pub changes: Vec<Change>,
+pub struct DiffWalker<'cb> {
+    pub cb: Box<dyn FnMut(Change) + 'cb>,
     pub lhs_root: RootSchema,
     pub rhs_root: RootSchema,
 }
 
-impl DiffWalker {
+impl<'cb> DiffWalker<'cb> {
+    pub fn new(
+        cb: Box<dyn FnMut(Change) + 'cb>,
```

Comment:
can we instead do:

```
struct DiffWalker<F: FnMut(Change)> {
    pub cb: F
}
```

I don't think we need a nameable DiffWalker type

the benefit of this is that the compiler can inline `cb` into the parent function, then it enables further optimization

Reply:
we would get an `reached the recursion limit while instantiating ...` error if we try to specify the type of callback as a generic parameter. This is because generic instantiation occurs infinitely in `diff_any_of`

Reply:
Can we just have `DiffWalker::count` method for saving memory? Although it is not as efficient as "callback" approach, the memory would be freed by the time `DiffWalker::count` is dropped and I think that is fair enough

Reply:
I didn't understand this comment at first, but you are right, there is a problem with generics. The reason is apparently that the closure contains the parent type in its signature, so the expansion of the type name recurses infinitely.

but we can still get rid of mandatory boxing, so I pushed a change that does that. Instead we can box only for the counter, and allow for inlining of the other callback

the super-optimized solution would define a new `Counter` struct that holds an integer and implements `FnMut` (or rather, a custom trait), but that is too annoying, even for me

Reply:
> we can still get rid of mandatory boxing

that was beyond my imagination. thanks

Reply:
> Can we just have private DiffWalker::count method for saving memory?

the problem is having a vector at all. if we have any vector and keep pushing stuff into it, it will use memory

the vector in its entirety is short-lived, so it is not like json-schema-diff will consume a lot of memory overall. but it will be slow because allocating, resizing, deallocating is slow.

_relatively_ slow, overall it will still be super fast. arguably this is premature optimization, but we have a `O(n^2)` loop here already to build the cross-product of all anyOf items between LHS and RHS, so I guess for some pathological input it could be a problem